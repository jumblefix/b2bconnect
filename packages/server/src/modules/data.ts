export const bookData = {
  title: 'My Book',
  coverImage: 'my-book.jpg',
  isbn: '1233434234',
  description:
    'Right now when I run npm test the output is just: ' +
    'and the number of passed and failed tests (when the tests are successful). ' +
    'I would like the output to include  and any other test descriptions' +
    ' (like how the output looks when an test is run).',
  rating: 0,
  price: 66.0,
  offerPrice: 56.0,
  categoryId: 'invalid',
  publishedYear: 1999,
};

export const user = {
  name: 'example',
  email: 'example@email.com',
  password: '123456',
};
