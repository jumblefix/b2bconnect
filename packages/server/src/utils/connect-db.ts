// tslint:disable-next-line:no-var-requires
const iconv = require('iconv-lite');
// tslint:disable-next-line:no-var-requires
const encodings = require('iconv-lite/encodings');
iconv.encodings = encodings;

import { createConnection, getConnectionOptions } from 'typeorm';

export const connectDb = async () => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);

  return createConnection({
    ...connectionOptions,
    name: 'default',
  });
};

export const connectDbTest = async (resetDB: boolean = false) => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);

  return createConnection({
    ...connectionOptions,
    name: 'default',
    synchronize: resetDB,
    dropSchema: resetDB,
  });
};
