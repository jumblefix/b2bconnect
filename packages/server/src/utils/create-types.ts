import { generateNamespace } from '@gql2ts/from-schema';
import * as fs from 'fs';
import * as path from 'path';
import { genSchema } from './schema-utils';

const typescriptTypes = generateNamespace('GQL', genSchema());

fs.writeFile(
  path.join(__dirname, '../typings/GQL/index.d.ts'),
  typescriptTypes,
  err => {
    console.error(err);
  },
);
