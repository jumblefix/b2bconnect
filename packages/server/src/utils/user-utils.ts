import { Request, Response } from 'express';
import * as IORedis from 'ioredis';
import { v4 } from 'uuid';
import { ValidationError } from 'yup';
import {
  FORGOT_PASSWORD_PREFIX,
  REDIS_SESSION_PREFIX,
  TokenTypes,
  USER_SESSION_PREFIX,
} from '../constants';
import { User } from '../entity/User';
import { GraphQLMiddlewareFunc, Resolver } from '../typings/app-utility-types';

export const createForgotPasswordLink = async (
  url: string,
  userId: string,
  redis: IORedis.Redis,
) => {
  const id = v4();
  await redis.set(`${FORGOT_PASSWORD_PREFIX}${id}`, userId, 'ex', 60 * 60);
  return `${url}/change-password/${id}`;
};

export const removeAllUsersSessions = async (
  userId: string,
  redis: IORedis.Redis,
) => {
  const sessionIds = await redis.lrange(
    `${USER_SESSION_PREFIX}${userId}`,
    0,
    -1,
  );

  const promises = [];
  // tslint:disable-next-line:prefer-for-of
  for (let i = 0; i < sessionIds.length; i += 1) {
    promises.push(redis.del(`${REDIS_SESSION_PREFIX}${sessionIds[i]}`));
  }
  await Promise.all(promises);
};

export const formatYupError = (err: ValidationError) => {
  const errors: [{path: string; message: string}] = [] as any;
  err.inner.forEach(e => {
    errors.push({
      path: e.path,
      message: e.message,
    });
  });

  return errors;
};

export const createMiddleware = (
  middlewareFunc: GraphQLMiddlewareFunc,
  resolverFunc: Resolver,
) => (parent: any, args: any, context: any, info: any) =>
  middlewareFunc(resolverFunc, parent, args, context, info);

export const middleware = async (
  resolver: Resolver,
  parent: any,
  args: any,
  context: any,
  info: any,
) => {
  return resolver(parent, args, context, info);
};

export const createTokenLink = async (
  url: string,
  userId: string,
  redis: IORedis.Redis,
  type: TokenTypes,
) => {
  const id = v4();
  await redis.set(id, userId, 'ex', 60 * 60 * 24);
  return `${url}/${type}/${id}`;
};

export const confirmEmail = async (req: Request, res: Response) => {
  const { id } = req.params;
  const RedisInstance = new IORedis();
  const userId = await RedisInstance.get(id);
  if (userId) {
    await User.update({ id: userId }, { confirmed: true });
    await RedisInstance.del(id);
    res.redirect(`${process.env.FRONTEND_HOST}/login`);
  } else {
    res.json({ message: 'invalid or expired register token' });
  }
};

export const resetPassword = async (req: Request, res: Response) => {
  const { id } = req.params;
  const RedisInstance = new IORedis();
  const userId = await RedisInstance.get(id);
  if (userId) {
    res.redirect(`${process.env.FRONTEND_HOST}/reset/${id}`);
  } else {
    res.json({ message: 'invalid or expired reset token' });
  }
};
