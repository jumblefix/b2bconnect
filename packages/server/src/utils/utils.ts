import {
  formatError as formatApolloError,
  isInstance as isApolloErrorInstance,
} from 'apollo-errors';
import * as nodemailer from 'nodemailer';
import slugify from 'slugify';
import { ValidationError } from 'yup';
import { Env } from '../constants';
import {
  AuthenticationError,
  AuthorizationError,
  InputValidationError,
} from './errors';

export const add = (...args: number[]) => args.reduce((c, p) => c + p);

export const formatYupError = (err: ValidationError) => {
  const errors: [{ path: string; message: string }] = [] as any;
  err.inner.forEach(e => {
    errors.push({
      path: e.path,
      message: e.message,
    });
  });
  return errors;
};

export function formatError(error: any) {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    // log internalData to stdout but not include it in the formattedError
    console.error(
      JSON.stringify({
        type: `error`,
        data: originalError.data,
        internalData: originalError.internalData,
      }),
    );
  }
  return formatApolloError(error);
}

export const makeSlug = (str: string) => slugify(str, { lower: true });

export const sendEmail = async ({
  to,
  subject,
  text,
  html,
}: SendEmailParams) => {
  let mailConfig: any;
  if (process.env.NODE_ENV === Env.production) {
    mailConfig = {
      host: 'smtp.sendgrid.net',
      port: 587,
      auth: {
        user: 'real.user',
        pass: 'verysecret',
      },
    };
  } else {
    mailConfig = {
      host: process.env.MAIL_HOST,
      port: process.env.MAIL_PORT,
      auth: {
        user: process.env.MAIL_USER,
        pass: process.env.MAIL_PASS,
      },
    };
  }

  const transport = nodemailer.createTransport(mailConfig);

  try {
    const info = await transport.sendMail({
      to,
      subject,
      text,
      html,
    });

    console.log(nodemailer.getTestMessageUrl(info));
  } catch (error) {
    console.error(error);
    throw new Error(error);
  }
  return true;
};

export const checkAdminRights = (session: any) => {
  const { userId, isAdmin } = session;
  if (!userId) {
    throw new AuthenticationError();
  }
  if (!isAdmin) {
    throw new AuthorizationError();
  }
};

export const validateInputs = async (schema: any, inputs: any) => {
  try {
    await schema.validate(inputs, { abortEarly: false });
  } catch (err) {
    const errors = formatYupError(err);
    throw new InputValidationError({
      data: errors,
    });
  }
};
