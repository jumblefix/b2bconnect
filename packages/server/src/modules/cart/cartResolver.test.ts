import { graphql } from 'graphql';
import * as ioredis from 'ioredis';
import { Connection } from 'typeorm';
import { createDb } from '../../utils/create-db';
import { genSchema } from '../../utils/schema-utils';
import { connectDbTest } from './../../utils/connect-db';
import { bookData, user } from './../data';

let connection: Connection;
let redis: ioredis.Redis;

const category = 'My Category';

beforeAll(async () => {
  await createDb();
  redis = new ioredis();
  connection = await connectDbTest(true);
});

afterAll(async () => {
  connection.close();
});

const getCartEmpty = {
  caseId: 'getCartEmpty',
  query: `{ getCart{ title } }`,
  session: { userId: '' },
};

const addCategory = {
  caseId: 'addCategory',
  query: `
       mutation {
         addCategory(name:"${category}"){
             id
             name
           }
         }`,
  session: {
    userId: '123',
    isAdmin: true,
  },
};

const addBook = {
  caseId: 'addBook',
  query: `
  mutation{
      addBook(
          title:"${bookData.title}",
          coverImage:"${bookData.coverImage}",
          isbn:"${bookData.isbn}",
          description:"${bookData.description}",
          rating:${bookData.rating},
          price:${bookData.price},
          offerPrice:${bookData.offerPrice},
          categoryId:"1",
          publishedYear:${bookData.publishedYear}
      ){
          id
          title
          isbn
          category{
              name
          }
      }
  }`,
  session: { userId: '' },
};

const registerUserValidData = {
  caseId: 'registerUserValidData',
  query: `mutation {
           register(name:"${user.name}", email: "${user.email}", password: "${
    user.password
  }") { email name id}}`,
  session: { userId: '' },
};

const addToCartQuery = (bookId: string) => `mutation {
  addToCart(bookId: "${bookId}") {
    id
    book {
      title
      slug
    }
    user {
      name
      email
    }
    title
  }
}
`;

const addToCart = (bid: string, _: any, uid: any) => {
  return {
    caseId: 'addToCart',
    query: addToCartQuery(bid),
    session: { userId: uid },
  };
};

const addToCartAgain = (bid: string, _: any, uid: any) => {
  return {
    caseId: 'addToCartAgain',
    query: addToCartQuery(bid),
    session: { userId: uid },
  };
};

const invalidBookIdQuery = addToCartQuery('123');

const addToCartQueryInvalidBook = (_: string, __: any, uid: any) => {
  return {
    caseId: 'addToCartQueryInvalidBook',
    query: invalidBookIdQuery,
    session: { userId: uid },
  };
};

const addToCartQueryInvalidBookWoLogin = (_: string, __: any, ___: any) => {
  return {
    caseId: 'addToCartQueryInvalidBook',
    query: invalidBookIdQuery,
    session: { userId: '' },
  };
};

const removeFromCart = (bid: string, _: any, uid: any) => {
  return {
    caseId: 'removeFromCart',
    query: `mutation{
               removeFromCart(bookId:"${bid}")
}`,
    session: { userId: uid },
  };
};

const removeWoLogin = {
  caseId: 'removeWoLogin',
  query: `mutation{
             removeFromCart(bookId:"123")
  }`,
  session: { userId: '' },
};

const removeWoLoginInvalidBook = (bid: string, _: any, uid: any) => {
  return {
    caseId: 'removeWoLogin',
    query: `mutation{
             removeFromCart(bookId:"${bid}")
  }`,
    session: { userId: uid },
  };
};

const emptyCartWoLogin = {
  caseId: 'emptyCartWoLogin',
  query: `mutation{emptyCart}`,
  session: { userId: '' },
};

const emptyCartWLogin = (_: string, __: any, uid: any) => {
  return {
    caseId: 'emptyCartWLogin',
    query: `mutation{emptyCart}`,
    session: { userId: uid },
  };
};

const getCart = (_: string, __: any, uid: any) => {
  return {
    caseId: 'getCart',
    query: `{getCart{ title }}`,
    session: { userId: uid },
  };
};

const addToCartLater = (bid: string, _: any, uid: any) => {
  return {
    caseId: 'addToCartLater',
    query: addToCartQuery(bid),
    session: { userId: uid },
  };
};

const emptyCartWLoginLater = (_: string, __: any, uid: any) => {
  return {
    caseId: 'emptyCartWLoginLater',
    query: `mutation{emptyCart}`,
    session: { userId: uid },
  };
};

describe('cart resolver', () => {
  const cases = [
    registerUserValidData,
    getCartEmpty,
    addCategory,
    addBook,
    addToCart,
    addToCartAgain,
    addToCartQueryInvalidBook,
    addToCartQueryInvalidBookWoLogin,
    removeFromCart,
    removeWoLogin,
    removeWoLoginInvalidBook,
    emptyCartWoLogin,
    emptyCartWLogin,
    getCart,
    addToCartLater,
    emptyCartWLoginLater,
  ];
  let bookId = '';
  let cid = '';
  let userId = '';

  cases.forEach(c => {
    const testCase: TestCase =
      typeof c === 'function' ? c(bookId, cid, userId) : c;
    it(`case: ${testCase.caseId}`, async () => {
      const { caseId, query, session } =
        typeof c === 'function' ? c(bookId, cid, userId) : c;
      const ctx = {
        redis,
        session,
        db: connection,
        url: 'http://localhost:4000',
        req: {},
      };
      const result = await graphql(genSchema(), query, null, ctx, {});
      if (caseId === 'addCategory') {
        cid = result.data!.addCategory.id;
        process.env.CATEGORY_ID = cid;
      }
      if (caseId === 'addBook') {
        bookId = result.data!.addBook.id;
        const bookTitle = result.data!.addBook.title;
        expect(bookTitle).toMatchSnapshot();
      } else if (caseId === 'registerUserValidData') {
        userId = result.data!.register.id;
      } else if (caseId === 'addToCart') {
        const cartTitle = result.data!.addToCart.title;
        expect(cartTitle).toMatchSnapshot();
      } else if (caseId === 'addToCartAgain') {
        expect(result.errors).toMatchSnapshot();
      } else if (caseId === 'addToCartLater') {
        expect(result.data!.addToCart.book).toMatchSnapshot();
      } else {
        expect(result).toMatchSnapshot();
      }
    });
  });
});
