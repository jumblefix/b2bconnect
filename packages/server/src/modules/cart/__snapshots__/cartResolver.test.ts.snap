// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart resolver case: addBook 1`] = `"My Book"`;

exports[`cart resolver case: addCategory 1`] = `
Object {
  "data": Object {
    "addCategory": Object {
      "id": "1",
      "name": "My Category",
    },
  },
}
`;

exports[`cart resolver case: addToCart 1`] = `"My Book"`;

exports[`cart resolver case: addToCartAgain 1`] = `
Array [
  [GraphQLError: Already in Cart],
]
`;

exports[`cart resolver case: addToCartLater 1`] = `
Object {
  "slug": "my-book",
  "title": "My Book",
}
`;

exports[`cart resolver case: addToCartQueryInvalidBook 1`] = `
Object {
  "data": Object {
    "addToCart": null,
  },
  "errors": Array [
    [GraphQLError: Item not found],
  ],
}
`;

exports[`cart resolver case: addToCartQueryInvalidBook 2`] = `
Object {
  "data": Object {
    "addToCart": null,
  },
  "errors": Array [
    [GraphQLError: Please login to continue],
  ],
}
`;

exports[`cart resolver case: emptyCartWLogin 1`] = `
Object {
  "data": Object {
    "emptyCart": null,
  },
  "errors": Array [
    [GraphQLError: Empty],
  ],
}
`;

exports[`cart resolver case: emptyCartWLoginLater 1`] = `
Object {
  "data": Object {
    "emptyCart": true,
  },
}
`;

exports[`cart resolver case: emptyCartWoLogin 1`] = `
Object {
  "data": Object {
    "emptyCart": null,
  },
  "errors": Array [
    [GraphQLError: Please login to continue],
  ],
}
`;

exports[`cart resolver case: getCart 1`] = `
Object {
  "data": Object {
    "getCart": Array [],
  },
}
`;

exports[`cart resolver case: getCartEmpty 1`] = `
Object {
  "data": Object {
    "getCart": null,
  },
  "errors": Array [
    [GraphQLError: Please login to continue],
  ],
}
`;

exports[`cart resolver case: removeFromCart 1`] = `
Object {
  "data": Object {
    "removeFromCart": true,
  },
}
`;

exports[`cart resolver case: removeWoLogin 1`] = `
Object {
  "data": Object {
    "removeFromCart": null,
  },
  "errors": Array [
    [GraphQLError: Please login to continue],
  ],
}
`;

exports[`cart resolver case: removeWoLogin 2`] = `
Object {
  "data": Object {
    "removeFromCart": null,
  },
  "errors": Array [
    [GraphQLError: Item not found],
  ],
}
`;
